x-pack/apm-server/sampling/eventstorage/prefix.go

```diff
@@ -0,0 +1,151 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package eventstorage
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/cockroachdb/pebble/v2"
+
+	"github.com/elastic/apm-data/model/modelpb"
+)
+
+const (
+	// NOTE(axw) these values (and their meanings) must remain stable
+	// over time, to avoid misinterpreting historical data.
+	entryMetaTraceSampled   byte = 's'
+	entryMetaTraceUnsampled byte = 'u'
+)
+
+var (
+	// ErrNotFound is returned by the RW.IsTraceSampled method,
+	// for non-existing trace IDs.
+	ErrNotFound = errors.New("key not found")
+)
+
+func NewPrefixReadWriter(db db, prefix byte, codec Codec) PrefixReadWriter {
+	return PrefixReadWriter{db: db, prefix: prefix, codec: codec}
+}
+
+type PrefixReadWriter struct {
+	db     db
+	prefix byte
+	codec  Codec
+}
+
+func (rw PrefixReadWriter) ReadTraceEvents(traceID string, out *modelpb.Batch) error {
+	var lb bytes.Buffer
+	lb.Grow(1 + len(traceID) + 1)
+	lb.WriteByte(rw.prefix)
+	lb.WriteString(traceID)
+	lb.WriteByte(':')
+
+	var ub bytes.Buffer
+	ub.Grow(lb.Len())
+	ub.Write(lb.Bytes()[:lb.Len()-1])
+	ub.WriteByte(';') // This is a hack to stop before next ID
```

Comment:
Maybe use a single buffer, and slice it for LowerBound and UpperBound?

Reply:
Actually, I don't think you need to specify the bounds at all if you're using SeekPrefixGE:

>> // SeekPrefixGE moves the iterator to the first key/value pair whose key is
>> // greater than or equal to the given key and which has the same "prefix" as
>> // the given key. The prefix for a key is determined by the user-defined
>> // Comparer.Split function. The iterator will not observe keys not matching the
>> // "prefix" of the search key.

Reply:
nice catch! Removed lb and ub from iter. @kruskall has some ideas on adding a fast path to not use a buffer for 16B trace ID, but that's out of scope of this PR.